File: app/src/main/java/com/fitaleks/heroesthesaurus/characters/MainActivityFragment.java
package com.fitaleks.heroesthesaurus.characters;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.fitaleks.heroesthesaurus.R;
import com.fitaleks.heroesthesaurus.data.Character;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static com.fitaleks.heroesthesaurus.util.Utils.checkNotNull;

public class MainActivityFragment extends Fragment implements CharactersContract.View {
    private static final String LOG_TAG = MainActivityFragment.class.getSimpleName();

    @BindView(R.id.recycler_view) RecyclerView recyclerView;

    private CharatersListAdapter mAdapter;
    private LinearLayoutManager mLinearLayoutManager;
    private boolean loading     = true;
    private int previousTotal   = 0;

    private CharactersContract.Presenter presenter;
    private Unbinder unbinder;

    public static MainActivityFragment newInstance() {
        return new MainActivityFragment();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.subscribe();
    }

    @Override
    public void onPause() {
        super.onPause();
        presenter.unsubscribe();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_main, container, false);
        unbinder = ButterKnife.bind(this, rootView);
//        recyclerView = (RecyclerView) rootView.findViewById(R.id.recycler_view);
        mAdapter = new CharatersListAdapter(new ArrayList<>());
        mLinearLayoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(mLinearLayoutManager);
        recyclerView.addOnScrollListener(listen);

        recyclerView.setAdapter(mAdapter);

        return rootView;
    }

    private int firstVisibleItem, visibleItemCount, totalItemCount;
    private final RecyclerView.OnScrollListener listen = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
        }

        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);
            visibleItemCount = recyclerView.getChildCount();
            totalItemCount = mLinearLayoutManager.getItemCount();
            firstVisibleItem = mLinearLayoutManager.findFirstVisibleItemPosition();

            if (loading) {
                if (totalItemCount > previousTotal) {
                    loading = false;
                    previousTotal = totalItemCount;
                }
            }
            if (!loading && (totalItemCount - visibleItemCount)
                    <= (firstVisibleItem + 2)) {
                // End has been reached

//                getObservable().observeOn(AndroidSchedulers.mainThread())
//                        .subscribe(MainActivityFragment.this::updateAdapter);

                loading = true;
            }
        }
    };

    @Override
    public void onDestroyView() {
        recyclerView.removeOnScrollListener(listen);
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void setLoadingIndicator(boolean active) {

    }

    @Override
    public void showCharacters(List<Character> characters) {
        mAdapter.addCharacters(characters);
    }

    @Override
    public void showLoadingError() {
        Toast.makeText(getContext(), "loading error", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void setPresenter(CharactersContract.Presenter presenter) {
        this.presenter = checkNotNull(presenter);
    }
}

File: app/src/main/java/com/fitaleks/heroesthesaurus/ui/MainActivityFragment.kt
package com.fitaleks.heroesthesaurus.ui

import android.arch.lifecycle.LifecycleFragment
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.fitaleks.heroesthesaurus.PATH_TO_ICONS
import com.fitaleks.heroesthesaurus.R
import com.fitaleks.heroesthesaurus.data.MtgSet
import com.fitaleks.heroesthesaurus.viewmodel.CharactersViewModel
import java.util.*

class MainActivityFragment : LifecycleFragment() {

    private val charactersRecyclerView: RecyclerView by lazy {
        view?.findViewById(R.id.recycler_view) as RecyclerView
    }
    private val swipeRefresh: SwipeRefreshLayout by lazy {
        view?.findViewById(R.id.swipe_refresh) as SwipeRefreshLayout
    }

    private var mAdapter: CharactersListAdapter = CharactersListAdapter(ArrayList<MtgSet>()).apply { setUseCharOfDay(true) }
    private var mLinearLayoutManager: LinearLayoutManager? = null
    private var loading = true
    private var previousTotal = 0

    private var firstVisibleItem: Int = 0
    private var visibleItemCount: Int = 0
    private var totalItemCount: Int = 0

    private val listen = object : RecyclerView.OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
        }

        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            visibleItemCount = recyclerView!!.childCount
            totalItemCount = mLinearLayoutManager!!.itemCount
            firstVisibleItem = mLinearLayoutManager!!.findFirstVisibleItemPosition()

            if (loading) {
                if (totalItemCount > previousTotal) {
                    loading = false
                    previousTotal = totalItemCount
                }
            }
            if (!loading && totalItemCount - visibleItemCount <= firstVisibleItem + 2) {
                // End has been reached

                //                getObservable().observeOn(AndroidSchedulers.mainThread())
                //                        .subscribe(MainActivityFragment.this::updateAdapter);

                loading = true
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_main, container, false)
    }

    private val TAG: String = MainActivityFragment::class.java.simpleName

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mLinearLayoutManager = LinearLayoutManager(activity)
        charactersRecyclerView.layoutManager = mLinearLayoutManager
        charactersRecyclerView.addOnScrollListener(listen)
        charactersRecyclerView.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        charactersRecyclerView.adapter = mAdapter
        swipeRefresh.isEnabled = false

        val appBarView: AppBarLayout? = activity.findViewById(R.id.appbar) as AppBarLayout
        charactersRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                if (charactersRecyclerView.computeVerticalScrollOffset() == 0) {
                    appBarView?.elevation = 0f
                } else {
                    appBarView?.elevation = resources.getDimension(R.dimen.appbar_elevation)
                }
            }
        })

//        val charactersViewModel = ViewModelProviders.of(this).get(CharactersViewModel::class.java)
//        charactersViewModel.getCharactersData()?.observe(this, android.arch.lifecycle.Observer { t -> t?.let { mAdapter.addCharacters(it) } })

        val setsViewModel = ViewModelProviders.of(this).get(CharactersViewModel::class.java)
        val listOfAllIcons = resources.assets.list(PATH_TO_ICONS)
        setsViewModel.getSets().observe(this, android.arch.lifecycle.Observer { t ->
            t?.filter { set ->
                if (!listOfAllIcons.contains(set.imageName())) {
                    Log.d(TAG, "${set.imageName()} has no icon")
                }

                set.type != "promo" && set.type != "reprint" && set.type != "box" && listOfAllIcons.contains(set.imageName())
            }
                    ?.let {
                        Log.d(TAG, "num of elems = ${it.size}")
                        mAdapter.addCharacters(it)
                    }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

    }

    override fun onDestroyView() {
        charactersRecyclerView.clearOnScrollListeners()
        super.onDestroyView()
    }

    companion object {
        fun newInstance(): MainActivityFragment {
            return MainActivityFragment()
        }
    }
}
